snippets/varia/heatmaply/heatmaply.Rmd

```diff
@@ -0,0 +1,214 @@
+summarizeHeatmap <- function(sce, expr_values, cluster_by, channels=NULL){
+  require(data.table)
+
+  # Argument checks
+  if(is.null(expr_values)){
+    expr_values <- names(assays(sce))[1]
+    print(paste0("Warning: Assay type not provided, '", expr_values, "' used."))
+  }
+
+  if(is.null(channels[1])){
+    channels <- rownames(sce)
+  }
+
+  if(!all(channels %in% rownames(sce))){
+    stop("Channel names do not correspond to the rownames of the assay")
+  }
+
+  if(is.null(cluster_by)){
+    stop("Cluster column not provided")
+  }
+
+  if(!cluster_by %in% colnames(colData(sce))){
+    stop("The cluster_by argument should correspond to a colData(sce) column")
+  }
+
+  # Convert the data to a melted format
+  dat <- as.data.table(t(assay(sce, expr_values)[channels,]))
+  dat[, id := colnames(sce)]
+  dat[, cluster := colData(sce)[, cluster_by]]
+  dat <- melt.data.table(dat, id.vars=c('id','cluster'), variable.name='channel', value.name=expr_values)
+
+  # Summarize the data
+  dat.summary <- dat[, list(
+    median.val = median(get(expr_values)),
+    mean.val = mean(get(expr_values)),
+    cellspercluster = .N),
+    by = c('channel', 'cluster')]
+
+  # Decast the summarized data and convert to a matrix. Currently only the median values are returned
+  hm.cell <- dcast.data.table(dat.summary, formula='cluster ~ channel', value.var='median.val')
+  hm.clusters <- hm.cell$cluster
+  hm.cell <- as.matrix(hm.cell[, -1, with=FALSE])
+  
+  # Add rownames
+  rownames(hm.cell) <- hm.clusters
+
+  # Return the median values
+  return(as.matrix(hm.cell))
+}
+```
+
+## Run the `summarizeHeatmap` function
+```{r run-summarizeHeatmap, message=FALSE}
+# Select the relevant channels
+good.channels <- rownames(sce)[! rownames(sce) %in% c('H3', 'PD-1', 'cPARP', 'Ir191', 'Ir193')]
+
+hm <- summarizeHeatmap(sce,
+                       expr_values = 'counts',
+                       cluster = 'CellType',
+                       channels = good.channels)
+```
+
+## Expression heatmap
+Displays normalized, scaled or percentized counts per cluster and channel  
+Normalize: https://cran.r-project.org/web/packages/heatmaply/vignettes/heatmaply.html#normalize  
+Scale: https://cran.r-project.org/web/packages/heatmaply/vignettes/heatmaply.html#scale  
+Percentize: https://cran.r-project.org/web/packages/heatmaply/vignettes/heatmaply.html#percentize  
+```{r expression-heatmap-normalize}
+# Normalize
+heatmaply(heatmaply::normalize(hm))
+
+# Scale
+heatmaply(hm, scale="column")
+
+# Percentize
+heatmaply(heatmaply::percentize(hm))
+```
+
+# Using asinh-transformed counts
+```{r}
+# Caculate transformed counts
+assay(sce, "exprs") <- asinh(counts(sce) / 1)
+
+# Summarize the data
+hm.exprs <- summarizeHeatmap(sce,
+                             expr_values = 'exprs',
+                             cluster = 'CellType',
+                             channels = good.channels)
+
+# Plot the heatmap
+heatmaply(heatmaply::normalize(hm.exprs))
+```
```

Comment:
I would suggest to also summarize the `counts` first and then arcsinh-transform them.